Analysis/WWCategorisation/src/WWCategorisationProcessor.cc

```diff
@@ -0,0 +1,431 @@
+    return;
+  }
+
+  // default category value: "other", works as cross check that nothing is forgotten
+  int true_cat = _n_cat-1;
+  int reco_cat = _n_cat-1;
+  
+  // get number of isolated leptons and photons -> used for basic categorisation
+  int n_electrons = col_electron->getNumberOfElements();
+  int n_muons = col_muon->getNumberOfElements();
+  int n_taus = col_tau->getNumberOfElements();
+  int n_photons = col_photons->getNumberOfElements();
+  int n_leptons = n_electrons + n_muons + n_taus;
+
+
+  // MCTruth categorisation
+
+  // particle 5 is - supposed to be! - the incoming electron/positron after ISR
+  MCParticle* mc = dynamic_cast <MCParticle*> (col_mcparticles->getElementAt(5));
+  int incoming_pdg = abs(mc->getPDG()); // get PDG of incoming e- or e+ after ISR
+
+  MCParticle* myelectron{}; // space for electron in sl channel for later check
+
+  // If the incoming particle is an electron/positron, check size of the daughters array; 4,3 or 1. If anything else: print
+  if(incoming_pdg == 11){
+    std::vector<MCParticle*> daughters = mc->getDaughters();
+    std::vector<MCParticle*> general_mcdaughters{}; // space for MC daughters
+    std::vector<int> general_daughters_pdgs = {}; // space for their PDGs
+
+    if (daughters.size() <= 3){ //if daughters are of length 3 or 2 
+      for (unsigned int i = 0; i < daughters.size(); ++i){ //go through all daughters and get their PDG
+        int daughters_pdg = abs(daughters[i]->getPDG());
+        
+        if (daughters_pdg > 20 && daughters_pdg < 30){ // make sure the incoming particle daughter is a boson
+          std::vector<MCParticle*> boson_daughters = daughters[i]->getDaughters();
+
+          for (unsigned int j = 0; j < boson_daughters.size(); ++j){ // go through the boson daughters and add them to daughters and pdg lists
+            general_mcdaughters.push_back(boson_daughters[j]);
+            general_daughters_pdgs.push_back(abs(boson_daughters.at(j)->getPDG()));
+          }
+        }
+        else{ //including remaining daughters that are not from W or Z
+          general_mcdaughters.push_back(daughters[i]); //including mc information of remaining daughters that are not from W or Z
+          general_daughters_pdgs.push_back(daughters_pdg);  //including PDGs of remaining daughters that are not from W or Z 
+        } //by the end we get all 4f daughters PDGs and all 4f daughters mc info
+      }
+    }
+    else if(daughters.size() == 4){ //go through all daughters of e-/e+ and save their PDG, then use the function again to count and categorise the events
+      for(unsigned int i = 0; i < daughters.size(); ++i){
+        general_mcdaughters.push_back(daughters[i]);
+        general_daughters_pdgs.push_back(abs(daughters[i]->getPDG()));
+      }
+    }else{ //PRINT OUT ACTUAL SIZE IF NOT ANY OF THE ONES MENTIONED
+      streamlog_out(DEBUG) << "atual size of daughters: " << daughters.size() << std::endl;
```

Comment:
Should this be a warning or an info message? I.e. is this something unusual or just something that doesn't fit this processor, but can happen regularly?

Reply:
Should not happen for signal events, but might for others, so MESSAGE/WARNING might be annoying if run on arbitrary samples.